### PR TITLE
create collection snapshot

### PR DESCRIPTION
**Pull request by nickprock**
# Description
When the embedder change qdrant create the collections snapshots and store it locally into `dormouse` folder.

<br>

dormouse

<br>

Related to issue #346 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: core/cat/memory/vector_memory.py [hunk 1]
```diff
@@ -129,7 +132,28 @@ def create_collection_if_not_exists(self):
                 log(f'Collection "{self.collection_name}" has the same embedder', "INFO")
             else:
                 log(f'Collection "{self.collection_name}" has different embedder', "WARNING")
-                # TODO: dump collection on disk before deleting, so it can be recovered
+                # dump collection on disk before deleting, so it can be recovered
+                # TODO: it needs a code refactoring
+
+                #dormouse = "dormouse/"
+                if os.path.isdir("dormouse/"):
```
Comment (pieroit) [inline core/cat/memory/vector_memory.py:139]: I propose to move the snapshot procedure in a separate method in the class

File: core/cat/memory/vector_memory.py [hunk 2]
```diff
@@ -129,7 +132,28 @@ def create_collection_if_not_exists(self):
...
+                    log(f'Directory dormouse exists', "INFO")
+                else:
+                    os.mkdir("dormouse/")
+
+                self.snapshot_info = self.client.create_snapshot(collection_name=self.collection_name)
+                # TODO: parametrize cat host and port
+                snapshot_url_in = "http://cheshire_cat_vector_memory:6333/collections/" + self.collection_name + "/snapshots/"+ self.snapshot_info.name
```
Comment (pieroit) [inline core/cat/memory/vector_memory.py:146]: snapshot should only happen if Qdrant is on another address, and not in [`:memory:` mode](https://github.com/cheshire-cat-ai/core/blob/f579b4badd942bb6862a055fcebb959db788d3c5/core/tests/conftest.py#L27) or [file mode](https://github.com/cheshire-cat-ai/core/blob/f579b4badd942bb6862a055fcebb959db788d3c5/core/cat/memory/vector_memory.py#L68).

There should be a way to ask directly to `self.client`: if host and port are present as attributes, we can snapshot:
- http client is in fact created by giving host and port [here](https://github.com/cheshire-cat-ai/core/blob/f579b4badd942bb6862a055fcebb959db788d3c5/core/cat/memory/vector_memory.py#L88)
- `self.client` should contain that information [as attributes](https://github.com/qdrant/qdrant-client/blob/730ed8d239d79bc1adec0d7c2efc31b63be9b8e5/qdrant_client/qdrant_client.py#L72)

**Review by pieroit — CHANGES_REQUESTED** [review]

**Comment by nickprock** [comment]
Hi @pieroit , I completely agree with you. The PR is work in progress.

File: core/cat/memory/vector_memory.py (hunk 2 above)
Comment (nickprock) [reply core/cat/memory/vector_memory.py]: Hi @pieroit I tried to print the attributes but

immagine

**Comment by nickprock** [comment]
@pieroit can you move the PR to another branch? 
We can use it as experiment

**Comment by pieroit** [comment]
@nickprock great idea will do (when I understand how) 
XD
Thanks

**Comment by nickprock** [comment]
> @nickprock great idea will do (when I understand how) XD Thanks

@pieroit  can use [this link](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request) or create a new branch and I edit the PR.

File: core/cat/memory/vector_memory.py (hunk 1 above)
Comment (nickprock) [reply core/cat/memory/vector_memory.py]: Done in core/cat/memory/vector_memory.py from row 253

File: core/cat/memory/vector_memory.py (hunk 2 above)
Comment (nickprock) [reply core/cat/memory/vector_memory.py]: Done in core/cat/memory/vector_memory.py from row 94 to 96

**Pull request merged**